include/irr/asset/CTexturePacker.h

```diff
@@ -81,57 +80,32 @@ class ITexturePacker
         }
         uint32_t x() const { return addr&X_MASK; }
         uint32_t y() const { return (addr>>(ADDR_LAYER_SHIFT>>1))&X_MASK; }
-        uint32_t layer() const { return addr>>ADDR_LAYER_SHIFT; }
+        uint32_t layer() const { return (addr>>ADDR_LAYER_SHIFT)&0xffffu; }
+
+        bool isValid() const { return (addr&0xffffu)==invalid_addr; }
+        bool hasMipTailAddr() const { return addr&(0xffffu<<16); }
+        STexOffset getMipTailAddr() const { return addr>>16; }
 
+        //upper 16 bits are used for address of mip-tail page
         uint32_t addr;
```

Comment:
just store it as `uint16_t[2]` to make life easier

Reply:
waaait.... `STexOffset` is a offset into virtual UV space (page table).... you don't store the `uint16_t` offset pair here... that goes into the page table!

Reply:
now page table and physical addr texture both have the same number of pages (texels in case of page tab) and layers, so STexOffset can represent both things. However ofc u're right that numbers of pages in physical addr tex and in page table could be completely independent

Reply:
so now i use STexOffset to represent page-wise address/offset into physical addr tex (what is stored in page table texels) AND texel-wise offset into page table (then upper 16 bits are unused ofc)

Reply:
> so now i use STexOffset to represent page-wise address/offset into physical addr tex (what is stored in page table texels) AND texel-wise offset into page table (then upper 16 bits are unused ofc)

Don't do that, its confusing. Separate types.

Reply:
> However ofc u're right that numbers of pages in physical addr tex and in page table could be completely independent

its useful to have a full 32bit Virtual Texture Offset, because then youre megatexture can be `2^(31 + pageSizeLog2*pageDimensions)` texels in size which becomes really fun  once streaming is implemented (page size of 1024 could allow a 64M x 32M texture which would be enough to store the whole of the Earth with resolution of 0.25m by 0.25m pixels)